### PR TITLE
ceph.in: Check return value when connecting

### PR DESCRIPTION
**Pull request by fullerdj**
The initial RADOS connection may time out. Detect this condition
and return an error message to the user instead of looping forever.

Signed-off-by: Douglas Fuller <dfuller@redhat.com>

### CODE REVIEW

File: src/ceph.in [hunk 1]
```diff
@@ -899,7 +899,11 @@ def main():
     try:
         if childargs and childargs[0] == 'ping':
             return ping_monitor(cluster_handle, childargs[1], timeout)
-        run_in_thread(cluster_handle.connect, timeout=timeout)
+        retval, outbuf, outs = run_in_thread(cluster_handle.connect,
+                                             timeout=timeout)
+        if retval == -errno.EINTR:
```
Comment (tchaikov) [inline src/ceph.in]: this could also caused by a real `KeyboardInterrupt`.

Comment (fullerdj) [inline src/ceph.in]: Good point. I adjusted the error message accordingly.

**Comment by alfredodeza** [comment]
@dmick these changes look ok to me, can you double check we aren't missing anything?

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-07-11_01:11:01-rados-wip-sage-testing-distro-basic-smithi did not go well:
<pre>
2017-07-11T01:17:02.138 INFO:teuthology.orchestra.run.smithi099:Running: 'sudo ceph --cluster ceph osd crush tunables default'
2017-07-11T01:17:02.190 INFO:tasks.ceph.mgr.x.smithi099.stderr:2017-07-11 01:17:02.194595 7f5ecc818380 -1 WARNING: all dangerous and experimental features are enabled.
2017-07-11T01:17:02.191 INFO:tasks.ceph.mgr.x.smithi099.stderr:2017-07-11 01:17:02.194740 7f5ecc818380 -1 WARNING: all dangerous and experimental features are enabled.
2017-07-11T01:17:02.191 INFO:tasks.ceph.mon.a.smithi099.stderr:2017-07-11 01:17:02.195949 7f839508ac00 -1 WARNING: all dangerous and experimental features are enabled.
2017-07-11T01:17:02.191 INFO:tasks.ceph.mon.a.smithi099.stderr:2017-07-11 01:17:02.196062 7f839508ac00 -1 WARNING: all dangerous and experimental features are enabled.
2017-07-11T01:17:02.191 INFO:tasks.ceph.mgr.x.smithi099.stderr:2017-07-11 01:17:02.196712 7f5ecc818380 -1 WARNING: all dangerous and experimental features are enabled.
2017-07-11T01:17:02.192 INFO:tasks.ceph.mon.a.smithi099.stderr:2017-07-11 01:17:02.198074 7f839508ac00 -1 WARNING: all dangerous and experimental features are enabled.
2017-07-11T01:17:02.236 INFO:teuthology.orchestra.run.smithi099.stderr:2017-07-11 01:17:02.241184 7feff4a1f700 -1 WARNING: all dangerous and experimental features are enabled.
2017-07-11T01:17:02.249 INFO:teuthology.orchestra.run.smithi099.stderr:2017-07-11 01:17:02.254776 7feff4a1f700 -1 WARNING: all dangerous and experimental features are enabled.
2017-07-11T01:17:03.112 INFO:teuthology.orchestra.run.smithi099.stderr:'NoneType' object is not iterable
2017-07-11T01:17:03.126 ERROR:teuthology.contextutil:Saw exception from nested tasks
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/contextutil.py", line 30, in nested
    vars.append(enter())
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/home/teuthworker/src/github.com_ceph_ceph-c_wip-sage-testing/qa/tasks/ceph.py", line 325, in crush_setup
    'osd', 'crush', 'tunables', profile])
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 193, in run
    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 414, in run
    r.wait()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 149, in wait
    self._raise_for_status()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 171, in _raise_for_status
    node=self.hostname, label=self.label
CommandFailedError: Command failed on smithi099 with status 1: 'sudo ceph --cluster ceph osd crush tunables default'
</pre>

**Comment by fullerdj** [comment]
Ok, @liewegas, I think this should fix the multiple ^C problem as well.

**Comment by tchaikov** [comment]
@fullerdj your PR includes some submodule changes by accident.

**Comment by fullerdj** [comment]
Oops, sorry. Fixed.

File: src/ceph.in [hunk 2]
```diff
@@ -899,7 +899,10 @@ def main():
     try:
         if childargs and childargs[0] == 'ping':
             return ping_monitor(cluster_handle, childargs[1], timeout)
-        run_in_thread(cluster_handle.connect, timeout=timeout)
+        result = run_in_thread(cluster_handle.connect, timeout=timeout)
+        if type(result) is tuple and result[0] == -errno.EINTR:
```
Comment (dmick) [inline src/ceph.in]: I don't know that any of these require fixing, but I do wonder what you think about the last point

Comment (fullerdj) [inline src/ceph.in]: I agree that the return value from run_in_thread is strange.

I'm not aware of the revision history here, but I think KeyboardInterrupt is very unlikely to be caught there since it's caught (and, strangely, raised so as to be caught) by run_in_thread (and ping_monitor primarily uses run_in_thread as well).

In fact, I think this issue is now more pervasive; squashing the KeyboardInterrupt leads to an infinite loop here and I suspect it may lead to bad behavior at other unchecked call sites.

Comment (dmick) [inline src/ceph.in]: I'd just as soon regularize the return from run_in_thread to a tuple, then, to reduce confusion and verbosity of test.

a long time ago my intent was to propagate the KeyboardInterrupt as the "the thread was interrupted either by human or timeout".  I don't know when that changed (could have been before I pushed originally) but it still seems like a plausible approach.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-07-12_19:30:01-rados-wip-sage-testing-distro-basic-smithi/

File: src/ceph.in (hunk 2 above)
Comment (dmick) [inline src/ceph.in]: or...not.

Comment (fullerdj) [inline src/ceph.in]: Sorry, it got merged before I changed it. I’ll do it in a separate PR.

Comment (fullerdj) [inline src/ceph.in]: or...not.

So, it turns out that run_in_thread can return darn near anything, depending on what you're running. So, it is not particularly amenable to standardization. We could have it return a dict with an "error" item, but otherwise we are stuck with this approach.
